File: app/src/main/java/com/bennet/wallet/activities/cards/ShowCardActivity.java
package com.bennet.wallet.activities.cards;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.ContextThemeWrapper;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.Space;
import android.widget.Toast;

import androidx.annotation.ColorInt;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bennet.wallet.preferences.AppPreferenceManager;
import com.bennet.wallet.preferences.CardPreferenceManager;
import com.bennet.wallet.services.DeleteCardService;
import com.bennet.wallet.R;
import com.bennet.wallet.adapters.ShowCardPropertyListItemAdapter;
import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.textview.MaterialTextView;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;

import com.bennet.wallet.utils.Utility.StringPair;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ShowCardActivity extends CardActivity {
    // UI
    protected ImageView cardCodeImageView;
    protected MaterialTextView cardCodePlainTextView;
    protected RecyclerView cardPropertiesView;
    protected View dividerCardProperties;
    protected View dividerCardImages;
    protected Space extraSpaceCardImages;

    // variables
    protected @ColorInt
    int codeForegroundColor;
    protected @ColorInt
    int codeBackgroundColor;
    protected List<StringPair> cardProperties = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_card);

        // hooks
        cardViewLayout = findViewById(R.id.show_card_card_view_layout);
        scrollView = findViewById(R.id.show_card_scroll_view);
        linearLayout = findViewById(R.id.show_card_linear_layout);
        cardCodeImageView = findViewById(R.id.show_card_code_image_view);
        cardCodePlainTextView = findViewById(R.id.show_card_code_plain_text);
        cardPropertiesView = findViewById(R.id.show_card_property_list_recycler_view);
        dividerCardProperties = findViewById(R.id.show_card_divider_card_properties);
        dividerCardImages = findViewById(R.id.show_card_divider_card_images);
        extraSpaceCardImages = findViewById(R.id.show_card_extra_space_card_images);

        // init
        initFromPreferences();

        // toolbar
        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(cardName);

        // variables
        codeForegroundColor = getResources().getColor(R.color.barcodeForegroundColor);
        codeBackgroundColor = getResources().getColor(R.color.barcodeBackgroundColor);

        // card properties recyclerview
        cardPropertiesView.setLayoutManager(new LinearLayoutManager(this));
        updateProperties();
        cardPropertiesView.setAdapter(new ShowCardPropertyListItemAdapter(cardProperties));

        init();

        // card view
        createCardView();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        scrollView.setScrollY(0); // refreshes scroll position

        cardView.removeFrontImage(); // hides old image, new image will be loaded later
        cardView.removeBackImage(); // hides old image, new image will be loaded later

        initFromPreferences();
        updateProperties();

        init();

        // card view
        updateFrontImage();
        updateBackImage();
    }

    protected void init() {
        // scroll view
        hideScrollbar();

        // card properties
        showCardProperties();
    }


    // handling action bar menu
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.show_card_or_password_action_bar_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.show_card_or_password_action_bar_edit) {
            Intent editIntent = new Intent(this, EditCardActivity.class);
            editIntent.putExtra(EditCardActivity.EXTRA_CARD_ID, ID);
            startActivity(editIntent);
            return true;
        }
        else if (itemId == R.id.show_card_or_password_action_bar_delete) {
            AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.RoundedCornersDialog));
            builder.setTitle(R.string.delete_card);
            builder.setMessage(R.string.delete_card_dialog_message);
            builder.setCancelable(true);
            builder.setPositiveButton(R.string.delete, (dialog, which) -> {
                finish(); // ALWAYS FINISH BEFORE STARTING OTHER ACTIVITY
                deleteCard();
                dialog.dismiss();
            });
            builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());
            builder.show();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return true;
    }


    // functions
    protected void deleteCard() {
        CardPreferenceManager.removeFromAllCardIDs(this, ID); // to make sure that HomeActivity doesn't display the card in the case DeleteCardService hasn't already finished deleting it

        Intent intent = new Intent(this, DeleteCardService.class);
        DeleteCardService.enqueueWork(this, intent, ID);
    }

    protected void showCardProperties() {
        // code
        if (!cardCode.equals("")) {

            if (!cardCodeTypeText) {
                cardCodePlainTextView.setVisibility(View.GONE);
            }
            else {
                cardCodePlainTextView.setVisibility(View.VISIBLE);
                cardCodePlainTextView.setText(cardCode);
            }

            cardCodeImageView.setVisibility(View.VISIBLE);
            Bitmap codeBitmap;
            try {
                if (codeIs1D(cardCodeType))
                    codeBitmap = createCode1D();
                else
                    codeBitmap = createCodeDefault();
            } catch (IllegalArgumentException | WriterException | ArrayIndexOutOfBoundsException e) {

                String detailedErrorMessage = "";
                if (AppPreferenceManager.isDetailedErrors(this))
                    detailedErrorMessage = System.getProperty("line.separator") + e.getLocalizedMessage();

                Toast.makeText(this, String.format(getString(R.string.show_card_visual_code_cannot_be_displayed), codeTypeIntToString(this, cardCodeType)) + detailedErrorMessage, Toast.LENGTH_LONG).show();
                cardCodeImageView.setVisibility(View.GONE);
                return;
            }
            cardCodeImageView.setImageBitmap(codeBitmap);
        } else {
            cardCodePlainTextView.setVisibility(View.GONE);
            cardCodeImageView.setVisibility(View.GONE);
        }

        // properties
        cardPropertiesView.getAdapter().notifyDataSetChanged(); // reload
        if (cardProperties.size() > 0 && !cardCode.equals(""))
            dividerCardProperties.setVisibility(View.VISIBLE);
        else
            dividerCardProperties.setVisibility(View.GONE);

        // images
        if ((currentFrontImage != null || currentBackImage != null) && (cardProperties.size() > 0 || !cardCode.equals(""))) {
            dividerCardImages.setVisibility(View.VISIBLE);
            if (cardProperties.size() > 0)
                extraSpaceCardImages.setVisibility(View.GONE);
            else
                extraSpaceCardImages.setVisibility(View.VISIBLE);
        }
        else
            dividerCardImages.setVisibility(View.GONE);
    }

    protected void updateProperties() {
        cardProperties.clear();

        if (!cardID.equals(""))
            cardProperties.add(new StringPair(getString(R.string.card_id), cardID));

        for (Integer propertyID : cardPropertyIDs) {
            String name = CardPreferenceManager.readCardPropertyName(this, ID, propertyID);
            String value = CardPreferenceManager.readCardPropertyValue(this, ID, propertyID);
            cardProperties.add(new StringPair(name, value));
        }
    }

    private Bitmap createCodeDefault() throws IllegalArgumentException, WriterException {
        final WidthAndHeight widthAndHeight = getBarcodeSize(cardCodeType);

        BitMatrix bitMatrix;
        bitMatrix = new MultiFormatWriter().encode(cardCode, cardCodeIntToBarcodeFormat(cardCodeType), widthAndHeight.width, widthAndHeight.height);

        final int bitMatrixWidth = bitMatrix.getWidth();
        final int bitMatrixHeight = bitMatrix.getHeight();

        int[] pixels = new int[bitMatrixWidth * bitMatrixHeight];
        for (int y = 0; y < bitMatrixHeight; y++) {
            final int offset = y * bitMatrixWidth;

            for (int x = 0; x < bitMatrixWidth; x++)
                pixels[offset + x] = bitMatrix.get(x, y) ? codeForegroundColor : codeBackgroundColor;
        }

        Bitmap bitmap = Bitmap.createBitmap(bitMatrixWidth, bitMatrixHeight, Bitmap.Config.ARGB_8888);
        bitmap.setPixels(pixels, 0, bitMatrixWidth, 0, 0, bitMatrixWidth, bitMatrixHeight);
        return bitmap;
    }

    private Bitmap createCode1D() throws IllegalArgumentException, WriterException {
        final WidthAndHeight widthAndHeight = getBarcodeSize(cardCodeType);

        BitMatrix bitMatrix;
        bitMatrix = new MultiFormatWriter().encode(cardCode, cardCodeIntToBarcodeFormat(cardCodeType), widthAndHeight.width, widthAndHeight.height, null);

        final int bitMatrixWidth = bitMatrix.getWidth();
        final int bitMatrixHeight = bitMatrix.getHeight();

        Bitmap bitmap = Bitmap.createBitmap(bitMatrixWidth, bitMatrixHeight, Bitmap.Config.ARGB_8888);
        for (int i = 0; i < bitMatrixWidth; i++) {
            int[] column = new int[bitMatrixHeight];
            Arrays.fill(column, bitMatrix.get(i, 0) ? codeForegroundColor : codeBackgroundColor);
            bitmap.setPixels(column, 0, 1, i, 0, 1, bitMatrixHeight);
        }
        return bitmap;
    }

    protected static class WidthAndHeight {
        int width;
        int height;

        public enum StdFormats {
            square,
            horizontal,
        }

        WidthAndHeight(float width, float height) {
            this.width = (int) width;
            this.height = (int) height;
        }

        WidthAndHeight(float layoutWidth, StdFormats format) {
            switch (format) {
                case square:
                    width = (int) (layoutWidth / 2);
                    //noinspection SuspiciousNameCombination
                    height = width;
                    break;
                case horizontal:
                    width = (int) layoutWidth;
                    height = width / 3;
                    break;
                default:
                    throw new IllegalStateException("Unexpected value: " + format);
            }
        }
    }

    private WidthAndHeight getBarcodeSize(int cardCodeType) {
        switch (cardCodeType) {
            case CardPreferenceManager.CARD_CODE_TYPE_AZTEC:
            case CardPreferenceManager.CARD_CODE_TYPE_QR:
            case CardPreferenceManager.CARD_CODE_TYPE_DATA_MATRIX:
                return new WidthAndHeight(getCalculatedLayoutWidth(), WidthAndHeight.StdFormats.square);

            case CardPreferenceManager.CARD_CODE_TYPE_PDF_417:
            case CardPreferenceManager.CARD_CODE_TYPE_CODABAR:
            case CardPreferenceManager.CARD_CODE_TYPE_CODE_39:
            case CardPreferenceManager.CARD_CODE_TYPE_CODE_93:
            case CardPreferenceManager.CARD_CODE_TYPE_CODE_128:
            case CardPreferenceManager.CARD_CODE_TYPE_EAN_8:
            case CardPreferenceManager.CARD_CODE_TYPE_EAN_13:
            case CardPreferenceManager.CARD_CODE_TYPE_ITF:
            case CardPreferenceManager.CARD_CODE_TYPE_UPC_A:
            case CardPreferenceManager.CARD_CODE_TYPE_UPC_E:
                return new WidthAndHeight(getCalculatedLayoutWidth(), WidthAndHeight.StdFormats.horizontal);

            default:
                throw new IllegalStateException("Unexpected value: " + cardCodeType);
        }
    }

}

File: app/src/main/java/com/izzdarki/wallet/ui/cards/ShowCardActivity.kt
package com.izzdarki.wallet.ui.cards

import android.content.DialogInterface
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.Space
import android.widget.Toast
import androidx.annotation.ColorInt
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import izzdarki.wallet.R
import com.izzdarki.wallet.ui.adapters.ShowPropertyAdapter
import com.izzdarki.wallet.preferences.AppPreferenceManager
import com.izzdarki.wallet.preferences.CardPreferenceManager
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.android.material.divider.MaterialDivider
import com.google.android.material.textview.MaterialTextView
import com.google.zxing.MultiFormatWriter
import com.google.zxing.WriterException
import com.google.zxing.common.BitMatrix
import java.util.*

class ShowCardActivity : CardActivity() {

    // region UI
    private lateinit var cardCodeImageView: ImageView
    private lateinit var cardCodePlainTextView: MaterialTextView
    private lateinit var cardPropertiesRecyclerView: RecyclerView
    private lateinit var dividerCardProperties: MaterialDivider
    private lateinit var dividerCardImages: MaterialDivider
    private lateinit var extraSpaceCardImages: Space
    private lateinit var labelsChipGroup: ChipGroup
    private lateinit var labelsDivider: MaterialDivider
    // endregion


    // region variables
    @ColorInt private var codeForegroundColor = 0
    @ColorInt private var codeBackgroundColor = 0
    // endregion


    // region lifecycle
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_show_card)

        // hooks
        cardViewLayout = findViewById(R.id.show_card_card_view_layout)
        scrollView = findViewById(R.id.show_card_scroll_view)
        linearLayout = findViewById(R.id.show_card_linear_layout)
        cardCodeImageView = findViewById(R.id.show_card_code_image_view)
        cardCodePlainTextView = findViewById(R.id.show_card_code_plain_text)
        cardPropertiesRecyclerView = findViewById(R.id.show_card_property_list_recycler_view)
        dividerCardProperties = findViewById(R.id.show_card_divider_card_properties)
        dividerCardImages = findViewById(R.id.show_card_divider_card_images)
        extraSpaceCardImages = findViewById(R.id.show_card_extra_space_card_images)
        labelsChipGroup = findViewById(R.id.labels_chip_group)
        labelsDivider = findViewById(R.id.labels_divider)

        // init
        initFromPreferences()

        // toolbar
        val toolbar = findViewById<MaterialToolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        // variables
        codeForegroundColor = resources.getColor(R.color.barcode_foreground_color)
        codeBackgroundColor = resources.getColor(R.color.barcode_background_color)

        // card properties recyclerview
        cardPropertiesRecyclerView.layoutManager = LinearLayoutManager(this)
        cardPropertiesRecyclerView.adapter = createShowPropertyAdapter()
        show()

        // password labels
        if (labels.isEmpty()) {
            labelsDivider.visibility = View.GONE
            labelsChipGroup.visibility = View.GONE
        } else {
            labelsDivider.visibility = View.VISIBLE
            labelsChipGroup.visibility = View.VISIBLE
            addLabelsToChipGroup()
        }

        // card view
        createCardView()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        scrollView.scrollY = 0 // refreshes scroll position
        cardView.removeFrontImage() // hides old image, new image will be loaded later
        cardView.removeBackImage() // hides old image, new image will be loaded later
        initFromPreferences()

        // hide/show labels chip group
        if (labels.isEmpty()) {
            labelsDivider.visibility = View.GONE
            labelsChipGroup.visibility = View.GONE
        } else {
            labelsDivider.visibility = View.VISIBLE
            labelsChipGroup.visibility = View.VISIBLE
            labelsChipGroup.removeAllViews()
            addLabelsToChipGroup()
        }

        // update
        cardPropertiesRecyclerView.adapter = createShowPropertyAdapter() // because cardProperties has been reassigned, a new adapter, that holds the new cardProperties is needed
        show()
        updateFrontImage()
        updateBackImage()
    }
    // endregion


    // region action bar
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.show_activity_action_bar_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId
        if (itemId == R.id.show_card_or_password_action_bar_edit) {
            val editIntent = Intent(this, EditCardActivity::class.java)
            editIntent.putExtra(EXTRA_CARD_ID, ID)
            startActivity(editIntent)
            return true
        } else if (itemId == R.id.show_card_or_password_action_bar_delete) {
            val builder =
                AlertDialog.Builder(this)
            builder.setTitle(R.string.delete_card)
            builder.setMessage(R.string.delete_card_dialog_message)
            builder.setCancelable(true)
            builder.setPositiveButton(R.string.delete) { dialog, _ ->
                finish() // ALWAYS FINISH BEFORE STARTING OTHER ACTIVITY
                CardPreferenceManager.removeComplete(this, ID)
                dialog.dismiss()
            }
            builder.setNegativeButton(
                android.R.string.cancel
            ) { dialog: DialogInterface, _: Int -> dialog.cancel() }
            builder.show()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return true
    }
    // endregion

    private fun show() {
        // hide scrollbar (TODO feature was removed, because it didn't do anything)
        //hideScrollbar()

        // card properties
        supportActionBar!!.title = cardName
        // code
        if (cardCode != "") {
            if (!cardCodeTypeText)
                cardCodePlainTextView.visibility = View.GONE
            else {
                cardCodePlainTextView.visibility = View.VISIBLE
                cardCodePlainTextView.text = cardCode
            }

            // lambda because there are 3 different exceptions to catch
            val catchExceptionFunc = { e: Exception ->
                val detailedErrorMessage =
                    if (AppPreferenceManager.isDetailedErrors(this))
                        System.getProperty("line.separator")!! + e.localizedMessage
                    else
                        ""
                Toast.makeText(
                    this,
                    String.format(
                        getString(R.string.show_card_visual_code_cannot_be_displayed),
                        codeTypeIntToString(this, cardCodeType)
                    ) + detailedErrorMessage,
                    Toast.LENGTH_LONG
                ).show()
                cardCodeImageView.visibility = View.GONE
            }

            cardCodeImageView.visibility = View.VISIBLE
            val codeBitmap: Bitmap
            try {
                codeBitmap = if (codeIs1D(cardCodeType)) createCode1D() else createCodeDefault()
            } catch (e: IllegalArgumentException) {
                catchExceptionFunc(e)
                return
            } catch (e: WriterException) {
                catchExceptionFunc(e)
                return
            } catch (e: ArrayIndexOutOfBoundsException) {
                catchExceptionFunc(e)
                return
            }
            cardCodeImageView.setImageBitmap(codeBitmap)
        } else {
            cardCodePlainTextView.visibility = View.GONE
            cardCodeImageView.visibility = View.GONE
        }

        // properties
        cardPropertiesRecyclerView.adapter?.notifyDataSetChanged() // reload
        if (cardProperties.size > 0 && cardCode != "")
            dividerCardProperties.visibility = View.VISIBLE
        else
            dividerCardProperties.visibility = View.GONE

        // images
        if ((currentFrontImage != null || currentBackImage != null) && (cardProperties.size > 0 || cardCode != "")) {
            dividerCardImages.visibility = View.VISIBLE
            if (cardProperties.size > 0)
                extraSpaceCardImages.visibility = View.GONE
            else
                extraSpaceCardImages.visibility = View.VISIBLE
        }
        else
            dividerCardImages.visibility = View.GONE
    }

    private fun addLabelsToChipGroup() {
        for (label in labels) {
            val chip = Chip(this)
            chip.text = label
            labelsChipGroup.addView(chip)
        }
    }

    private fun createShowPropertyAdapter(): ShowPropertyAdapter {
        return ShowPropertyAdapter(cardProperties) {
            // hide all secret values in adapter each time the user presses any of the visibility buttons
            for (position in cardProperties.indices) {
                val holder = cardPropertiesRecyclerView.findViewHolderForAdapterPosition(position) as? ShowPropertyAdapter.ViewHolder
                holder?.setValueHidden(cardProperties[position].secret)
            }
        }
    }

    @Throws(IllegalArgumentException::class, WriterException::class)
    private fun createCodeDefault(): Bitmap {
        val widthAndHeight = getBarcodeSize(cardCodeType)
        val bitMatrix: BitMatrix = MultiFormatWriter().encode(
            cardCode,
            cardCodeIntToBarcodeFormat(cardCodeType),
            widthAndHeight.width,
            widthAndHeight.height
        )
        val bitMatrixWidth = bitMatrix.width
        val bitMatrixHeight = bitMatrix.height
        val pixels = IntArray(bitMatrixWidth * bitMatrixHeight)
        for (y in 0 until bitMatrixHeight) {
            val offset = y * bitMatrixWidth
            for (x in 0 until bitMatrixWidth) pixels[offset + x] =
                if (bitMatrix[x, y]) codeForegroundColor else codeBackgroundColor
        }
        val bitmap = Bitmap.createBitmap(bitMatrixWidth, bitMatrixHeight, Bitmap.Config.ARGB_8888)
        bitmap.setPixels(pixels, 0, bitMatrixWidth, 0, 0, bitMatrixWidth, bitMatrixHeight)
        return bitmap
    }

    @Throws(IllegalArgumentException::class, WriterException::class)
    private fun createCode1D(): Bitmap {
        val widthAndHeight = getBarcodeSize(cardCodeType)
        val bitMatrix: BitMatrix = MultiFormatWriter().encode(
            cardCode,
            cardCodeIntToBarcodeFormat(cardCodeType),
            widthAndHeight.width,
            widthAndHeight.height,
            null
        )
        val bitMatrixWidth = bitMatrix.width
        val bitMatrixHeight = bitMatrix.height
        val bitmap = Bitmap.createBitmap(bitMatrixWidth, bitMatrixHeight, Bitmap.Config.ARGB_8888)
        for (i in 0 until bitMatrixWidth) {
            val column = IntArray(bitMatrixHeight)
            Arrays.fill(column, if (bitMatrix[i, 0]) codeForegroundColor else codeBackgroundColor)
            bitmap.setPixels(column, 0, 1, i, 0, 1, bitMatrixHeight)
        }
        return bitmap
    }

    private class WidthAndHeight {
        var width = 0
        var height = 0

        enum class StdFormats {
            Square, Horizontal
        }

        constructor(width: Float, height: Float) {
            this.width = width.toInt()
            this.height = height.toInt()
        }

        constructor(layoutWidth: Float, format: StdFormats) {
            when (format) {
                StdFormats.Square -> {
                    width = (layoutWidth / 2).toInt()
                    height = width
                }
                StdFormats.Horizontal -> {
                    width = layoutWidth.toInt()
                    height = width / 3
                }
            }
        }
    }

    private fun getBarcodeSize(cardCodeType: Int): WidthAndHeight {
        return when (cardCodeType) {
            CardPreferenceManager.CARD_CODE_TYPE_AZTEC,
            CardPreferenceManager.CARD_CODE_TYPE_QR,
            CardPreferenceManager.CARD_CODE_TYPE_DATA_MATRIX
                -> WidthAndHeight(calculatedLayoutWidth, WidthAndHeight.StdFormats.Square)

            CardPreferenceManager.CARD_CODE_TYPE_PDF_417,
            CardPreferenceManager.CARD_CODE_TYPE_CODABAR,
            CardPreferenceManager.CARD_CODE_TYPE_CODE_39,
            CardPreferenceManager.CARD_CODE_TYPE_CODE_93,
            CardPreferenceManager.CARD_CODE_TYPE_CODE_128,
            CardPreferenceManager.CARD_CODE_TYPE_EAN_8,
            CardPreferenceManager.CARD_CODE_TYPE_EAN_13,
            CardPreferenceManager.CARD_CODE_TYPE_ITF,
            CardPreferenceManager.CARD_CODE_TYPE_UPC_A,
            CardPreferenceManager.CARD_CODE_TYPE_UPC_E
                -> WidthAndHeight(calculatedLayoutWidth, WidthAndHeight.StdFormats.Horizontal)

            else -> throw IllegalStateException("Unexpected value: $cardCodeType")
        }
    }
}